File: ios/PdfPageImageModule.swift
import ExpoModulesCore
import Foundation
import PDFKit
import UIKit

private enum PdfPageImageError: CodedError {
  case cleanupFailed(String)
  case emptyDocument
  case fileNotFound(String)
  case invalidUri(String)
  case openFailed(String)
  case renderFailed(String)
  case writeFailed(String)

  var code: String {
    switch self {
    case .cleanupFailed:
      return "ERR_PDF_CLEANUP_FAILED"
    case .emptyDocument:
      return "ERR_PDF_EMPTY_DOCUMENT"
    case .fileNotFound:
      return "ERR_PDF_FILE_NOT_FOUND"
    case .invalidUri:
      return "ERR_PDF_INVALID_URI"
    case .openFailed:
      return "ERR_PDF_OPEN_FAILED"
    case .renderFailed:
      return "ERR_PDF_RENDER_FAILED"
    case .writeFailed:
      return "ERR_PDF_WRITE_FAILED"
    }
  }

  var description: String {
    switch self {
    case .cleanupFailed(let message),
         .fileNotFound(let message),
         .invalidUri(let message),
         .openFailed(let message),
         .renderFailed(let message),
         .writeFailed(let message):
      return message
    case .emptyDocument:
      return "PDF document does not contain any pages."
    }
  }
}

public final class PdfPageImageModule: Module {
  private let outputDirectoryName = "pdf-page-image"

  public func definition() -> ModuleDefinition {
    Name("PdfPageImage")

    AsyncFunction("generateAllPages") { (uri: String, scale: Double) in
      return try self.generateAllPages(uri: uri, scale: scale)
    }

    AsyncFunction("cleanupPages") { (uris: [String]) in
      try self.cleanupPages(uris: uris)
    }
  }

  private func generateAllPages(uri: String, scale: Double) throws -> [[String: Any]] {
    let fileUrl = try normalizePdfUrl(uri)

    guard FileManager.default.fileExists(atPath: fileUrl.path) else {
      throw PdfPageImageError.fileNotFound("PDF file not found at path: \(fileUrl.path)")
    }

    guard let document = PDFDocument(url: fileUrl) else {
      throw PdfPageImageError.openFailed("Unable to open PDF document at path: \(fileUrl.path)")
    }

    guard document.pageCount > 0 else {
      throw PdfPageImageError.emptyDocument
    }

    try ensureOutputDirectoryExists()

    return try (0..<document.pageCount).map { index in
      guard let page = document.page(at: index) else {
        throw PdfPageImageError.renderFailed("Unable to access PDF page at index \(index).")
      }

      return try renderPage(page, scale: scale)
    }
  }

  private func cleanupPages(uris: [String]) throws {
    let outputDirectory = try outputDirectory()
    var cleanupErrors: [String] = []

    for uri in Set(uris) {
      guard let fileUrl = normalizeOutputUrl(uri) else {
        continue
      }

      guard fileUrl.path.hasPrefix(outputDirectory.path) else {
        continue
      }

      if !FileManager.default.fileExists(atPath: fileUrl.path) {
        continue
      }

      do {
        try FileManager.default.removeItem(at: fileUrl)
      } catch {
        cleanupErrors.append("Failed to remove \(fileUrl.lastPathComponent): \(error.localizedDescription)")
      }
    }

    if !cleanupErrors.isEmpty {
      throw PdfPageImageError.cleanupFailed(cleanupErrors.joined(separator: "\n"))
    }
  }

  private func renderPage(_ page: PDFPage, scale: Double) throws -> [String: Any] {
    let pageBounds = normalizedBounds(for: page)
    let renderScale = max(scale, 1)
    let renderSize = CGSize(
      width: max(pageBounds.width * renderScale, 1),
      height: max(pageBounds.height * renderScale, 1)
    )

    let format = UIGraphicsImageRendererFormat.default()
    format.scale = 1
    let renderer = UIGraphicsImageRenderer(size: renderSize, format: format)

    let image = renderer.image { context in
      UIColor.white.setFill()
      context.fill(CGRect(origin: .zero, size: renderSize))

      context.cgContext.saveGState()
      context.cgContext.translateBy(x: 0, y: renderSize.height)
      context.cgContext.scaleBy(x: renderScale, y: -renderScale)
      page.draw(with: .mediaBox, to: context.cgContext)
      context.cgContext.restoreGState()
    }

    guard let data = image.pngData() else {
      throw PdfPageImageError.writeFailed("Failed to encode rendered PDF page as PNG.")
    }

    let outputUrl = try outputDirectory().appendingPathComponent("\(UUID().uuidString).png")

    do {
      try data.write(to: outputUrl, options: .atomic)
    } catch {
      throw PdfPageImageError.writeFailed("Failed to write rendered PDF page: \(error.localizedDescription)")
    }

    return [
      "height": Int(renderSize.height.rounded()),
      "uri": outputUrl.absoluteString,
      "width": Int(renderSize.width.rounded())
    ]
  }

  private func normalizedBounds(for page: PDFPage) -> CGRect {
    let bounds = page.bounds(for: .mediaBox)
    let rotation = ((page.rotation % 360) + 360) % 360

    if rotation == 90 || rotation == 270 {
      return CGRect(x: bounds.origin.x, y: bounds.origin.y, width: bounds.height, height: bounds.width)
    }

    return bounds
  }

  private func normalizePdfUrl(_ uri: String) throws -> URL {
    let trimmedUri = uri.trimmingCharacters(in: .whitespacesAndNewlines)

    guard !trimmedUri.isEmpty else {
      throw PdfPageImageError.invalidUri("PDF URI is empty.")
    }

    if trimmedUri.hasPrefix("file://") {
      guard let fileUrl = URL(string: trimmedUri), fileUrl.isFileURL else {
        throw PdfPageImageError.invalidUri("Invalid file URI: \(trimmedUri)")
      }
      return fileUrl
    }

    if trimmedUri.hasPrefix("/") {
      return URL(fileURLWithPath: trimmedUri)
    }

    throw PdfPageImageError.invalidUri("Unsupported PDF URI: \(trimmedUri)")
  }

  private func normalizeOutputUrl(_ uri: String) -> URL? {
    let trimmedUri = uri.trimmingCharacters(in: .whitespacesAndNewlines)

    if trimmedUri.hasPrefix("file://") {
      return URL(string: trimmedUri)
    }

    if trimmedUri.hasPrefix("/") {
      return URL(fileURLWithPath: trimmedUri)
    }

    return nil
  }

  private func outputDirectory() throws -> URL {
    return FileManager.default.temporaryDirectory
      .appendingPathComponent(outputDirectoryName, isDirectory: true)
  }

  private func ensureOutputDirectoryExists() throws {
    let directory = try outputDirectory()

    if FileManager.default.fileExists(atPath: directory.path) {
      return
    }

    do {
      try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    } catch {
      throw PdfPageImageError.writeFailed("Failed to create PDF page output directory: \(error.localizedDescription)")
    }
  }
}

File: android/src/main/java/expo/modules/pdfpageimage/PdfPageImageModule.kt
package expo.modules.pdfpageimage

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Matrix
import android.graphics.pdf.PdfRenderer
import android.net.Uri
import android.os.ParcelFileDescriptor
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.util.UUID

private const val OUTPUT_DIR_NAME = "pdf-page-image"

private class PdfInvalidUri(message: String) : CodedException("ERR_PDF_INVALID_URI", message, null)

private class PdfFileNotFound(message: String) : CodedException("ERR_PDF_FILE_NOT_FOUND", message, null)

private class PdfOpenFailed(message: String) : CodedException("ERR_PDF_OPEN_FAILED", message, null)

private class PdfEmptyDocument : CodedException("ERR_PDF_EMPTY_DOCUMENT", "PDF document does not contain any pages.", null)

private class PdfRenderFailed(message: String) : CodedException("ERR_PDF_RENDER_FAILED", message, null)

private class PdfWriteFailed(message: String) : CodedException("ERR_PDF_WRITE_FAILED", message, null)

private class PdfCleanupFailed(message: String) : CodedException("ERR_PDF_CLEANUP_FAILED", message, null)

class PdfPageImageModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  override fun definition() = ModuleDefinition {
    Name("PdfPageImage")

    AsyncFunction("generateAllPages") { uri: String, scale: Double ->
      generateAllPages(uri, scale)
    }

    AsyncFunction("cleanupPages") { uris: List<String> ->
      cleanupPages(uris)
    }
  }

  private fun outputDirectory(): File {
    val dir = File(context.cacheDir, OUTPUT_DIR_NAME)
    if (!dir.exists()) {
      if (!dir.mkdirs()) {
        throw PdfWriteFailed("Failed to create PDF page output directory: ${dir.path}")
      }
    }
    return dir
  }

  private fun parsePdfUri(uriString: String): Uri {
    val trimmed = uriString.trim()
    if (trimmed.isEmpty()) {
      throw PdfInvalidUri("PDF URI is empty.")
    }
    return when {
      trimmed.startsWith("content://") -> Uri.parse(trimmed)
      trimmed.startsWith("file://") -> Uri.parse(trimmed)
      trimmed.startsWith("/") ->
        Uri.parse(File(trimmed).absoluteFile.toURI().toString())
      else -> throw PdfInvalidUri("Unsupported PDF URI: $trimmed")
    }
  }

  private fun openReadOnlyPfd(uri: Uri): ParcelFileDescriptor {
    return when (uri.scheme?.lowercase()) {
      "content" -> {
        try {
          context.contentResolver.openFileDescriptor(uri, "r")
        } catch (e: SecurityException) {
          throw PdfOpenFailed("Unable to open PDF from content URI: ${e.localizedMessage}")
        } catch (e: FileNotFoundException) {
          throw PdfFileNotFound("PDF file not found for content URI: ${e.localizedMessage}")
        } ?: throw PdfOpenFailed("Unable to open PDF from content URI.")
      }
      "file" -> {
        val path = uri.path ?: throw PdfInvalidUri("Invalid file URI: $uri")
        val file = File(path)
        if (!file.exists()) {
          throw PdfFileNotFound("PDF file not found at path: ${file.path}")
        }
        try {
          ParcelFileDescriptor.open(file, ParcelFileDescriptor.MODE_READ_ONLY)
        } catch (e: Exception) {
          throw PdfOpenFailed("Unable to open PDF document at path: ${file.path} (${e.localizedMessage})")
        }
      }
      else -> throw PdfInvalidUri("Unsupported PDF URI scheme: ${uri.scheme}")
    }
  }

  private fun generateAllPages(uriString: String, scale: Double): List<Map<String, Any>> {
    val uri = parsePdfUri(uriString)
    val pfd = openReadOnlyPfd(uri)
    val renderer = try {
      PdfRenderer(pfd)
    } catch (e: Exception) {
      pfd.close()
      throw PdfOpenFailed("Unable to open PDF document: ${e.localizedMessage}")
    }

    return try {
      if (renderer.pageCount <= 0) {
        throw PdfEmptyDocument()
      }

      val outputDir = outputDirectory()
      val renderScale = scale.coerceAtLeast(1.0).toFloat()
      val results = ArrayList<Map<String, Any>>(renderer.pageCount)

      for (i in 0 until renderer.pageCount) {
        val page = try {
          renderer.openPage(i)
        } catch (e: Exception) {
          throw PdfRenderFailed("Unable to access PDF page at index $i: ${e.localizedMessage}")
        }

        try {
          val baseW = page.width.coerceAtLeast(1)
          val baseH = page.height.coerceAtLeast(1)
          val outW = (baseW * renderScale).toInt().coerceAtLeast(1)
          val outH = (baseH * renderScale).toInt().coerceAtLeast(1)

          val bitmap = Bitmap.createBitmap(outW, outH, Bitmap.Config.ARGB_8888)
          val canvas = Canvas(bitmap)
          canvas.drawColor(Color.WHITE)

          val matrix = Matrix()
          matrix.setScale(renderScale, renderScale)

          try {
            page.render(bitmap, null, matrix, PdfRenderer.Page.RENDER_MODE_FOR_DISPLAY)
          } catch (e: Exception) {
            bitmap.recycle()
            throw PdfRenderFailed("Failed to render PDF page at index $i: ${e.localizedMessage}")
          }

          val outFile = File(outputDir, "${UUID.randomUUID()}.png")
          try {
            FileOutputStream(outFile).use { fos ->
              if (!bitmap.compress(Bitmap.CompressFormat.PNG, 100, fos)) {
                throw PdfWriteFailed("Failed to encode rendered PDF page as PNG.")
              }
            }
          } catch (e: PdfWriteFailed) {
            bitmap.recycle()
            throw e
          } catch (e: Exception) {
            bitmap.recycle()
            throw PdfWriteFailed("Failed to write rendered PDF page: ${e.localizedMessage}")
          }

          bitmap.recycle()

          results.add(
            mapOf(
              "width" to outW,
              "height" to outH,
              "uri" to outFile.absoluteFile.toURI().toString()
            )
          )
        } finally {
          page.close()
        }
      }

      results
    } finally {
      renderer.close()
      pfd.close()
    }
  }

  private fun cleanupPages(uris: List<String>) {
    val outputDir = outputDirectory().canonicalFile
    val outputPrefix = outputDir.absolutePath + File.separator
    val cleanupErrors = mutableListOf<String>()

    for (raw in uris.toSet()) {
      val file = resolveCleanupFile(raw.trim()) ?: continue
      val canonical: String
      try {
        canonical = file.canonicalPath
      } catch (_: Exception) {
        continue
      }

      if (!canonical.startsWith(outputPrefix)) {
        continue
      }

      if (!file.exists()) {
        continue
      }

      try {
        if (!file.delete()) {
          cleanupErrors.add("Failed to remove ${file.name}: delete returned false")
        }
      } catch (e: Exception) {
        cleanupErrors.add("Failed to remove ${file.name}: ${e.localizedMessage}")
      }
    }

    if (cleanupErrors.isNotEmpty()) {
      throw PdfCleanupFailed(cleanupErrors.joinToString("\n"))
    }
  }

  private fun resolveCleanupFile(uriString: String): File? {
    if (uriString.isEmpty()) {
      return null
    }
    return when {
      uriString.startsWith("content://") -> null
      uriString.startsWith("file://") -> {
        val uri = Uri.parse(uriString)
        val path = uri.path ?: return null
        File(path)
      }
      uriString.startsWith("/") -> File(uriString)
      else -> null
    }
  }
}
